zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,329 @@
+    /// Sapling note commitment tree.
+    fn new_leaf(block: Arc<Block>, network: Network, sapling_root: &sapling::tree::Root) -> Self {
+        let node_data = block_to_history_node(block, network, sapling_root);
+        let low_entry: zcash_history::Entry = node_data.into();
+        low_entry.into()
+    }
+
+    /// Create a node (non-leaf) Entry from the encoded node data and the indices of
+    /// its children (in the array representation of the MMR tree).
+    fn new_node(
+        branch_id: ConsensusBranchId,
+        data: NodeData,
+        left_idx: u32,
+        right_idx: u32,
+    ) -> Result<Self, io::Error> {
+        let node_data = zcash_history::NodeData::from_bytes(branch_id.into(), data.encoded)?;
+        let low_entry = zcash_history::Entry::new(
+            node_data,
+            zcash_history::EntryLink::Stored(left_idx),
+            zcash_history::EntryLink::Stored(right_idx),
+        );
+        Ok(low_entry.into())
+    }
+}
+
+impl Tree {
+    /// Create a MMR tree with the given length.
+    ///
+    /// The `peaks` are the peaks of the MMR tree to build and their position in the
+    /// array representation of the tree.
+    /// The `extra` are extra nodes that enable removing nodes from the tree, and their position.
+    ///
+    /// Note that the length is usually larger than the length of `peaks` and `extra`, since
+    /// you don't need to pass every node, just the peaks of the tree (plus extra).
+    fn new(
+        network: Network,
+        network_upgrade: NetworkUpgrade,
+        length: u32,
+        peaks: &HashMap<u32, &Entry>,
+        extra: &HashMap<u32, &Entry>,
+    ) -> Result<Self, io::Error> {
+        let branch_id = network_upgrade
+            .branch_id()
+            .expect("unexpected pre-Overwinter MMR history tree");
+        let mut peaks_vec = Vec::new();
+        for (idx, entry) in peaks {
+            let low_entry = zcash_history::Entry::from_bytes(branch_id.into(), entry.encoded)?;
+            peaks_vec.push((*idx, low_entry));
+        }
+        let mut extra_vec = Vec::new();
+        for (idx, entry) in extra {
+            let low_entry = zcash_history::Entry::from_bytes(branch_id.into(), entry.encoded)?;
+            extra_vec.push((*idx, low_entry));
+        }
+        let tree = zcash_history::Tree::new(length, peaks_vec, extra_vec);
+        Ok(Tree { network, tree })
+    }
+
+    /// Append a new block to the tree, as a new leaf.
+    ///
+    /// `sapling_root` is the root of the Sapling note commitment tree of the block.
+    ///
+    /// Returns a vector of nodes added to the tree (leaf + internal nodes).
+    fn append_leaf(
+        &mut self,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+    ) -> Result<Vec<NodeData>, zcash_history::Error> {
+        let node_data = block_to_history_node(block, self.network, sapling_root);
+        let appended = self.tree.append_leaf(node_data)?;
```

Comment:
Each history tree is only allowed to have blocks from a single network upgrade.

Can we panic if other Zebra code tries to append a node that's not in the current network upgrade?
(We'd need to store the network upgrade in our `Tree`, and compare it with the network upgrade we look up from the stored network and the new block's height.)

That way, we catch this kind of bug early, before we corrupt our stored data.

In the finalized state, we'll store history by height. That will completely prevent this kind of error, because block heights are unique. (See #2134.) But we still have to guard against it in the non-finalized state.